From 1049d6f78ad4cf422c86744d5c2009aab0cf78a7 Mon Sep 17 00:00:00 2001
From: David Howells <dhowells@redhat.com>
Date: Tue, 25 Nov 2014 16:33:51 -0500
Subject: [fs] fuse: restructure ->rename2()

Message-id: <20141125163351.18115.32652.stgit@warthog.procyon.org.uk>
Patchwork-id: 100911
O-Subject: [PATCH 13/47] fuse: restructure ->rename2()
Bugzilla: 985875
RH-Acked-by: Steve Best <sbest@redhat.com>

From: Miklos Szeredi <mszeredi@suse.cz>

Make ->rename2() universal, i.e. able to handle zero flags.  This is to
make future change of the API easier.

Signed-off-by: Miklos Szeredi <mszeredi@suse.cz>
Original-commit: 4237ba43b65aa989674c89fc4f2fe46eebc501ee
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/fuse/dir.c b/fs/fuse/dir.c
index bc6243d..8f378c2 100644
--- a/fs/fuse/dir.c
+++ b/fs/fuse/dir.c
@@ -818,13 +818,6 @@ static int fuse_rename_common(struct inode *olddir, struct dentry *oldent,
  return err;
 }
 
-static int fuse_rename(struct inode *olddir, struct dentry *oldent,
-         struct inode *newdir, struct dentry *newent)
-{
- return fuse_rename_common(olddir, oldent, newdir, newent, 0,
-      FUSE_RENAME, sizeof(struct fuse_rename_in));
-}
-
 static int fuse_rename2(struct inode *olddir, struct dentry *oldent,
    struct inode *newdir, struct dentry *newent,
    unsigned int flags)
@@ -835,17 +828,30 @@ static int fuse_rename2(struct inode *olddir, struct dentry *oldent,
  if (flags & ~(RENAME_NOREPLACE | RENAME_EXCHANGE))
   return -EINVAL;
 
- if (fc->no_rename2 || fc->minor < 23)
-  return -EINVAL;
+ if (flags) {
+  if (fc->no_rename2 || fc->minor < 23)
+   return -EINVAL;
 
- err = fuse_rename_common(olddir, oldent, newdir, newent, flags,
-     FUSE_RENAME2, sizeof(struct fuse_rename2_in));
- if (err == -ENOSYS) {
-  fc->no_rename2 = 1;
-  err = -EINVAL;
+  err = fuse_rename_common(olddir, oldent, newdir, newent, flags,
+      FUSE_RENAME2,
+      sizeof(struct fuse_rename2_in));
+  if (err == -ENOSYS) {
+   fc->no_rename2 = 1;
+   err = -EINVAL;
+  }
+ } else {
+  err = fuse_rename_common(olddir, oldent, newdir, newent, 0,
+      FUSE_RENAME,
+      sizeof(struct fuse_rename_in));
  }
+
  return err;
+}
 
+static int fuse_rename(struct inode *olddir, struct dentry *oldent,
+         struct inode *newdir, struct dentry *newent)
+{
+ return fuse_rename2(olddir, oldent, newdir, newent, 0);
 }
 
 static int fuse_link(struct dentry *entry, struct inode *newdir,
-- 
1.7.1